drivers/clock_control/clock_control_arm_scmi.c

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright 2024 NXP
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/drivers/firmware/scmi/clk.h>
+#include <zephyr/drivers/clock_control.h>
+#include <zephyr/logging/log.h>
+
+LOG_MODULE_REGISTER(arm_scmi_clock);
+
+#define DT_DRV_COMPAT arm_scmi_clock
+
+struct scmi_clock_data {
+	uint32_t clk_num;
+};
+
+static int scmi_clock_on_off(const struct device *dev,
+			     clock_control_subsys_t clk, bool on)
+{
+	struct scmi_clock_data *data;
+	struct scmi_protocol *proto;
+	uint32_t clk_id;
+	struct scmi_clock_config cfg;
+
+	proto = dev->data;
+	data = proto->data;
+	clk_id = POINTER_TO_UINT(clk);
+
+	if (clk_id >= data->clk_num) {
+		return -EINVAL;
+	}
+
+	memset(&cfg, 0, sizeof(cfg));
+
+	cfg.attributes = SCMI_CLK_CONFIG_ENABLE_DISABLE(on);
+	cfg.clk_id = clk_id;
+
+	return scmi_clock_config_set(proto, &cfg);
+}
+
+static int scmi_clock_on(const struct device *dev, clock_control_subsys_t clk)
+{
+	return scmi_clock_on_off(dev, clk, true);
+}
+
+static int scmi_clock_off(const struct device *dev, clock_control_subsys_t clk)
+{
+	return scmi_clock_on_off(dev, clk, false);
+}
+
+static int scmi_clock_get_rate(const struct device *dev,
+			       clock_control_subsys_t clk, uint32_t *rate)
+{
+	struct scmi_clock_data *data;
+	struct scmi_protocol *proto;
+	uint32_t clk_id;
+
+	proto = dev->data;
+	data = proto->data;
+	clk_id = POINTER_TO_UINT(clk);
+
+	if (clk_id >= data->clk_num) {
+		return -EINVAL;
+	}
+
+	return scmi_clock_rate_get(proto, clk_id, rate);
+}
+
+static struct clock_control_driver_api scmi_clock_api = {
+	.on = scmi_clock_on,
+	.off = scmi_clock_off,
+	.get_rate = scmi_clock_get_rate,
+};
```

Comment:
Is there any plan to implement set_rate? I think that may be much complex to consider parent clock.
But that should be requirement.
Thanks.

Reply:
There was no plan to add this in the initial support. I tested this on 95 with some basic samples (i.e: `hello_world` and `synchronization`) in which `set_rate()` was not needed so this is why this wasn't included here. Of course, this feature is important so it should be added at some point. Do we need to have this feature in the initial support or is it fine if we add it later on?

Reply:
Sure, it's ok for me to have it added in the future :)